src/main/java/org/killbill/billing/plugin/bridge/api/resolver/remote/RemoteResolver.java

```diff
@@ -37,20 +38,44 @@ public RemoteResolver(final KillBillClient client, final RequestOptions requestO
     }
 
     // TODO We could execute some requests in parallel -- exercise for the reader...
-    public RemoteResolverResponse resolve(final RemoteResolverRequest request) {
+    public RemoteResolverResponse resolve(final RemoteResolverRequest request) throws KillBillClientException {
         final RemoteResolverResponse.RemoteResolverResponseBuilder result = new RemoteResolverResponse.RemoteResolverResponseBuilder();
         request.getRequests()
-                .stream()
-                .forEachOrdered(r -> {
-                    try {
-                        r.resolve(client, requestOptions, result);
-                        logger.info(String.format("RemoteResolver [%d]: type='%s', id='%s' -> resolvedId='%s'",
-                                request.getRequestId(), r.getType(), r.getSrcKey(), result.getMapping(r.getType())));
-                    } catch (final KillBillClientException e) {
-                        logger.warn(String.format("RemoteResolver failed to execute request='%s', id='%s'", r.getType(), r.getSrcKey()));
-                    }
-                });
-
+               .stream()
+               .forEachOrdered(r -> {
+                   boolean resolved = false;
+                   WrappedKillBillClientException clientException = null;
+                   try {
+                       r.resolve(client, requestOptions, result);
+                       resolved = true;
+                   } catch (final KillBillClientException e) {
+                       logger.warn(String.format("RemoteResolver [%d]: KillBillClientException...", request.getRequestId()), e);
```

Comment:
No need for `String.format` (`{}` notation).

Reply:
I fixed this one, but before i got further and fix them all, a few questions:
* Is there a limit in the # of '{}' i can have ?
* This seems to work when using `logger.xx` but not when using `logService.log()` (`OSGIKillbillLogService`), which brings me to following question: what is the right way to log from a plugin?

Reply:
> * Is there a limit in the # of '{}' i can have ?

I don't think so, see the implementation [here](https://github.com/qos-ch/slf4j/blob/5e93462dd781677fb75aefc6cc7df65822970ca0/slf4j-api/src/main/java/org/slf4j/helpers/MessageFormatter.java#L186-L241).

> * This seems to work when using `logger.xx` but not when using `logService.log()` (`OSGIKillbillLogService`)

Indeed: `{}` is part of the SLF4J API. `OSGIKillbillLogService` implements `LogService` (OSGI spec), which only understands already-formatted messages.

> which brings me to following question: what is the right way to log from a plugin?  

Short answer: use SLF4J directly.

Longer answer: we need to use the OSGI mechanism `LogService` in plugins (e.g. to give us access to `LogEntry` events displayed in Kaui) but because the API is somewhat limited (e.g. no formatting options) and because we want logs to respect the main `logback.xml` (to support obfuscation for instance), we have a bridge in place to redirect `SLF4J` logs from plugins into the OSGI `LogService`and back into `SLF4J` in Kill Bill (https://github.com/killbill/killbill/wiki/Logback-recipes#handling-plugin-logs).